CI: Test NumPy against OpenBLAS weekly builds

honno opened the pull request:
This PR tests NumPy against OpenBLAS weekly builds, as well as stable builds like before.

Things to note:

* `linux_meson.yml` is updated to use the pre-existing `cibw_before_build.sh`/`openblas_support.py` scripts. These scripts are updated to support the weekly OpenBLAS builds, mirroring the changes from https://github.com/scipy/scipy/pull/18889 (testing SciPy against OpenBLAS weeklies). It feels weird to almost perfectly duplicate these scripts, and not use things like `getopt`, but it looks like they should be made redundant in the future anywho (https://github.com/scipy/scipy/pull/18889#discussion_r1265031999).

* Before `linux_meson.yml` got OpenBLAS via `apt-get install libopenblas-serial-dev`. This PR currently pulls OpenBLAS from [`multibuild-wheels-staging`](https://anaconda.org/multibuild-wheels-staging/openblas-libs/)... maybe there's value in testing the Ubuntu package too?

* ~We're hoping to get these stable and weekly builds that live in `multibuild-wheels-staging` to `scientific-python-nightly-wheels` https://github.com/MacPython/openblas-libs/pull/111, which looks like it could happen very soon.~ That PR got merged, so we can wait on it updating to make sure things work but should be good to go. See also https://github.com/scipy/scipy/pull/18942

cc @mattip @steppi @rgommers

honno committed:
MAINT: Use `cibw_before_build.sh` in `linux_meson.yml`

honno committed:
TST: Test with nightly OpenBLAS builds

rgommers commented:
@honno the failures look relevant.

steppi commented on tools/openblas_support.py:
```diff
@@ -15,8 +15,9 @@
 OPENBLAS_LONG = 'v0.3.23-246-g3d31191b'
-BASE_LOC = 'https://anaconda.org/multibuild-wheels-staging/openblas-libs'
-BASEURL = f'{BASE_LOC}/{OPENBLAS_LONG}/download'
+BASE_LOC = (
+    'https://anaconda.org/scientific-python-nightly-wheels/openblas-libs'
+)
```
`BASE_LOC`  should be different depending on whether or not the weekly is used. If the weekly isn't being used, the url should still be `'https://anaconda.org/multibuild-wheels-staging/openblas-libs'`. 
 
In https://github.com/scipy/scipy/pull/18942 `base_loc` as a keyword arg to `download_openblas`. See [here](https://github.com/scipy/scipy/blob/9300bee6608a637b3a16b9ef5fe14256cd9a99fd/tools/openblas_support.py#L18-L21).

mattip replied:
> If the weekly isn't being used, the url should still be 'https://anaconda.org/multibuild-wheels-staging/openblas-libs'

I think this is because the tag needed by SciPy, `v0.3.20-571-g3dec11c6` has not been uploaded to https://anaconda.org/scientific-python-nightly-wheels/openblas-libs. But for NumPy, the tag `v0.3.23-246-g3d31191b` __has__ been uploaded to the new location, so here we can use the single site.

mattip replied:
ahh, the travis job did not upload, so indeed the new site is missing the aarch64 tarballs.

steppi replied:
I see. I think I misunderstood. From the name, I thought that `scientific-python-nightly-wheels` is only for nightlies and that tagged releases are still being uploaded to the old location.

mattip replied:
I restarted the travis job, let's see if this time it succeeds (the token was missing).

honno replied:
So yeah to emphasise, `scientific-python-nightly-wheels/openblas-libs` is now holding the stable and weekly builds.

mattip commented:
The macos-arm64 [wheel build failure](https://github.com/numpy/numpy/pull/24250/checks?check_run_id=15290989191) is due to the new code, it seems the `getopts` parsing is not right on macOS.

steppi commented:
> The macos-arm64 [wheel build failure](https://github.com/numpy/numpy/pull/24250/checks?check_run_id=15290989191) is due to the new code, it seems the `getopts` parsing is not right on macOS.

Right, this is why I reverted using `getopt` in the SciPy version of this PR  https://github.com/scipy/scipy/pull/18889#issuecomment-1645957704

honno commented:
> > The macos-arm64 [wheel build failure](https://github.com/numpy/numpy/pull/24250/checks?check_run_id=15290989191) is due to the new code, it seems the `getopts` parsing is not right on macOS.
> 
> Right, this is why I reverted using `getopt` in the SciPy version of this PR [scipy/scipy#18889 (comment)](https://github.com/scipy/scipy/pull/18889#issuecomment-1645957704)

Fixed, all green now :sweat_smile:

mattip commented on .github/workflows/linux_meson.yml:
```diff
@@ -35,7 +40,14 @@ jobs:
+          sudo RUNNER_OS=${{ runner.os }} tools/wheels/cibw_before_build.sh .
+        else
+          sudo RUNNER_OS=${{ runner.os }} tools/wheels/cibw_before_build.sh .
+        fi
```
What is the difference between the two stanzas and why is the extra `chmod` now needed?

Also: this is meant to only install OpenBLAS. But sourcing `cibw_before_build.sh` does more
- installs a gfortran on macOS
- mangles the license file

Maybe you could copy the 3 needed lines (on linux) here directly, without calling a script. You would need to conditionally add a `--nightly` argument to the first line 
```
if ... 
  target=$(python tools/openblas_support.py --nightly)
else
  ...
fi
sudo cp -r $target/lib/* /usr/lib
sudo cp $target/include/* /usr/include
```

honno replied:
I had forgot the missing `--nightly` flag, but yes your solution seems nicer anywho.

mattip commented on tools/openblas_support.py:
```diff
@@ -156,6 +158,10 @@ def setup_openblas(plat=get_plat(), ilp64=get_ilp64()):
+    openblas_version = "HEAD" if nightly else OPENBLAS_LONG
+    typ = download_openblas(
+        tmp, plat, ilp64, openblas_version=openblas_version
+    )
```
I think this logic belongs in `download_openblas()`, not here

honno replied:
I might be a bit fuzzy brained, but I think this is fine? Just I had duplicated the `typ` assignment, which I've now removed.

mattip replied:
My review relates to the construction of openblas_version here, rather than inside `download_openblas`. To me it makes more sense to do all the URL construction and logic in one place, and not have a piece here and a piece inside `download_openblas`. So I would prefer the signature of `download_openblas` get a `nightly=False` kwarg instead of an `openblas_version` kwarg.

honno replied:
Ah I see makes sense, updated.

mattip commented on .github/workflows/linux_meson.yml:
```diff
@@ -53,5 +65,6 @@ jobs:
       env:
         TERM: xterm-256color
       run: |
+        export LD_LIBRARY_PATH="/usr/local/lib/"  # to find libopenblas.so.0
```
Move this to the general `env:` stanza for this CI job (line 31), I think building should use it as well to detect OpenBLAS.

mattip commented on tools/openblas_support.py:
```diff
@@ -120,7 +121,8 @@ def download_openblas(target, plat, ilp64):
         return None
-    filename = f'{BASEURL}/openblas{fnsuffix}-{OPENBLAS_LONG}-{suffix}'
+    BASEURL = f'{BASE_LOC}/{openblas_version}/download'
+    filename = f'{BASEURL}/openblas{fnsuffix}-{openblas_version}-{suffix}'
```
Nit: As you noticed, BASEURL is not really a global (capitalized) variable, this should be styled differently now, and the `BASEURL` all along should have had a `_` between the words. Maybe get rid of it altogether?

honno replied:
Removed it altogether.

honno committed:
BUG: Add missing `--nightly` flag

honno committed:
MAINT: Specify `LD_LIBRARY_PATH` as a workflow env

honno committed:
MAINT: Remove intermediate `BASEURL` variable

honno committed:
MAINT: Remove first and redundant `typ` assignment

honno committed:
MAINT: Don't use `cibw_before_build.sh` in `linux_meson.yml`

honno reviewed (commented):
Per Matti's suggestions I've simplified a few things... and we should actually be testing against the weekly builds of OpenBLAS now :sweat_smile:

honno committed:
MAINT: Construct `openblas_version` in `download_openblas()` instead

mattip commented:
Two last things: 
- the [jobs are called](https://github.com/numpy/numpy/actions/runs/5659523027/job/15333255047?pr=24250) "meson_spin (true)" and "meson_spin (false)". Could you think of a way to get better names?
- Is there a check you can point to that the correct OpenBLAS version is being used? The azure builds use this, which might need to grow a `--nightly` option that would verify that the OpenBLAS version is __not__ the one in the ``OPENBLAS_LONG` value: https://github.com/numpy/numpy/blob/511f77d1202b5f23ca7e8988d9455e803d5303d6/azure-pipelines.yml#L180-L185

honno committed:
MAINT: Prettify `linux_meson.yml` workflow name

honno commented:
> the [jobs are called](https://github.com/numpy/numpy/actions/runs/5659523027/job/15333255047?pr=24250) "meson_spin (true)" and "meson_spin (false)". Could you think of a way to get better names?

Yep, so TIL `name` attribute in jobs determines the... name, and we can do a hacky condition in a GitHub expression (see https://github.com/actions/runner/issues/409#issuecomment-727565588) to change it depending on say matrix values.

> Is there a check you can point to that the correct OpenBLAS version is being used?

I played around with things, no success so far. Notably `openblas_support.py::test_version()` uses `import numpy`, which gives us an `ImportError`, seemingly by virtue of how NumPy is built in this workflow. Also its a pain to explore/debug how `test_version()` works locally. Might write a separate script for now that just prints the version.

mattip commented:
> Notably openblas_support.py::test_version() uses import numpy, which gives us an ImportError, seemingly by virtue of how NumPy is built in this workflow.

Since this is using the `spin` workflow, you will need to use `spin python ...` to execute the check

honno commented:
> which might need to grow a `--nightly` option that would verify that the OpenBLAS version is **not** the one in the `OPENBLAS_LONG` value:

The OpenBLAS nightly build seemingly has the same "version" as the stable build, looking at this [run](https://github.com/honno/numpy/actions/runs/5669150635/job/15362521364#step:5:761)

```
ValueError: nightly OpenBLAS version should not be 0.3.23.dev, got [{'user_api': 'blas', 'internal_api': 'openblas', 'num_threads': 2, 'prefix': 'libopenblas', 'filepath': '/usr/lib/libopenblasp-r0.3.23.dev.so', 'version': '0.3.23.dev', 'threading_layer': 'pthreads', 'architecture': 'SkylakeX'}]
```

I checked against `OPENBLAS_V`, as thats what `openblas_support.py::test_version()` uses—I might of misunderstand you though. I assume the version attribute (however `threadpoolctl` gets it) would be whatever the latest stable version is anywho, as its not something updated. I'm not sure what to look for really...

Reproducing building with nightly OpenBLAS builds locally has been a huge time sink already, been hard to experiment :upside_down_face:

rgommers commented:
That `test_version` check is horrible, I suggest we ignore it. Looking at the CI logs, it's clear that the "stable" job does `python tools/openblas_support.py` while the "nightly" job does `python tools/openblas_support.py --nightly`. If we'd just uncomment a `print` statement in `download_openblas` to verify the exact artifact we're downloading, we should be all good here.

honno committed:
MAINT: print filename in `openblas_support.py::download_openblas()`

honno commented:
> If we'd just uncomment a `print` statement in `download_openblas` to verify the exact artifact we're downloading, we should be all good here.

Introduced a new comment as wasn't sure what to do with `# print(f"Downloading {length} from {filename}", file=sys.stderr)`. Looks like we're downloading the right files!

https://github.com/numpy/numpy/actions/runs/5716123665/job/15486923504?pr=24250#step:4:63
https://github.com/numpy/numpy/actions/runs/5716123665/job/15486923255?pr=24250#step:4:63

rgommers committed:
Merge branch 'main' into test-openblas-nightlies

rgommers commented:
The Circle CI build wasn't touched in this PR and is only failing because this branch is behind. The Cirrus CI failures are unrelated too. Still, a few too many failures to confidently merge - let's retry that now (I merged `main` into this PR).

rgommers commented:
Okay, the `intel_spr_sde_test` failures are clearly unrelated, and the rest all looks good now. In it goes. Thanks @honno, @mattip & @steppi!
